src/mount-support.c

```diff
@@ -416,6 +241,257 @@ static void sc_setup_mount_profiles(const char *security_tag)
+	if (mount(config->rootfs_dir, scratch_dir, NULL, MS_REC | MS_BIND, NULL)
+	    < 0) {
+		die("cannot perform operation: mount --rbind %s %s",
+		    config->rootfs_dir, scratch_dir);
+	}
+	// Make the scratch directory recursively private. Nothing done there will
+	// be shared with any peer group, This effectively detaches us from the
+	// original namespace and coupled with pivot_root below serves as the
+	// foundation of the mount sandbox.
+	debug("performing operation: mount --make-rslave %s", scratch_dir);
+	if (mount("none", scratch_dir, NULL, MS_REC | MS_SLAVE, NULL) < 0) {
+		die("cannot perform operation: mount --make-rslave %s",
+		    scratch_dir);
+	}
+	// Bind mount certain directories from the host filesystem to the scratch
+	// directory. Mount events will propagate in both into and out of the
+	// peer group. This way the running application can alter any global state
+	// visible on the host and in other snaps.
+	for (const char **path = config->bidirectional_mounts; *path != NULL;
+	     path++) {
+		must_snprintf(dst, sizeof dst, "%s/%s", scratch_dir, *path);
+		debug("performing operation: mount --rbind %s %s", *path, dst);
+		if (mount(*path, dst, NULL, MS_REC | MS_BIND, NULL) < 0) {
+			die("cannot perform operation: mount --rbind %s %s",
+			    *path, dst);
+		}
+	}
+	// Bind mount certain directories from the host filesystem to the scratch directory.
+	// Mount events will only propagate inwards to the namespace. This way the running
+	// application cannot alter any global state apart from that of its own snap.
+	for (const char **path = config->unidirectional_mounts; *path != NULL;
+	     path++) {
+		must_snprintf(dst, sizeof dst, "%s%s", scratch_dir, *path);
+		debug("performing operation: mount --rbind %s %s", *path, dst);
+		if (mount(*path, dst, NULL, MS_REC | MS_BIND, NULL) != 0) {
+			die("cannot perform operation: mount --rbind %s %s",
+			    *path, dst);
+		}
+		debug("performing operation: mount --make-rslave %s", dst);
+		if (mount("none", dst, NULL, MS_REC | MS_SLAVE, NULL) != 0) {
+			die("cannot perform operation: mount --make-rslave %s",
+			    dst);
+		}
+	}
+	// Since we mounted /etc from the host filesystem to the scratch directory,
+	// we may need to put /etc/alternatives from the desired root filesystem
+	// (e.g. the core snap) back. This way the behavior of running snaps is not
+	// affected by the alternatives directory from the host, if one exists.
+	//
+	// https://bugs.launchpad.net/snap-confine/+bug/1580018
+	const char *etc_alternatives = "/etc/alternatives";
+	if (access(etc_alternatives, F_OK) == 0) {
+		must_snprintf(src, sizeof src, "%s%s", config->rootfs_dir,
+			      etc_alternatives);
+		must_snprintf(dst, sizeof dst, "%s%s", scratch_dir,
+			      etc_alternatives);
+		// NOTE: MS_SLAVE so that the started process cannot maliciously mount
+		// anything into those places and affect the system on the outside.
+		debug("performing operation: mount --bind -o slave %s %s", src,
```

Comment:
I think "-o slave" works but we should change "-o slave" to "--make-slave" to match the rest of the debugging output.